coins/src/utils/coins3/produce.ts

```diff
@@ -15,102 +18,170 @@ export type Dynamo = {
-        chain,
-        source: adapter,
-        redirect: redirectPid,
-      };
+      return; // for coins-metadata, we now rely on getMetadataRecord elsewhere
     case "coins-current":
-      return { pid, price, confidence, source: adapter, mcap };
+      return { pid, price, confidence, adapter, mcap, updateTs: Math.floor(Date.now() / 1000) };
     case "coins-timeseries":
-      return { pid, price, confidence, source: adapter, ts: SK, mcap };
+      return { pid, price, confidence, adapter, ts: SK, mcap };
     default:
       throw new Error(`Topic '${topic}' is not valid`);
   }
 }
-function splitPk(pk: string): { chain: string; address: string; pid: string } {
-  const assetPrefix: string = "asset#";
-  const coingeckoPrefix: string = "coingecko#";
-
-  if (pk.toLowerCase().startsWith(coingeckoPrefix)) {
-    const address = pk.substring(coingeckoPrefix.length).toLowerCase();
-    return {
-      chain: "coingecko",
-      address,
-      pid: `coingecko:${address}`,
-    };
+
+function splitPk(pk: string, decimals?: number): { chain?: string; address?: string; pid: string } {
+  const pid = normalizeCoinId(pk);
+  const record: any = { pid };
+  if (pid.includes(":")) {
+    record.chain = pid.split(":")[0];
+    record.address = pid.slice(record.chain.length + 1);
+  } else if (pid.length === 42 && pid.startsWith("0x")) {
+    record.chain = "ethereum";
+    record.address = pid;
+  } else if (!decimals && !pid.startsWith("0x")) {
+    record.chain = "coingecko";
+  }
+  return record;
+}
+
+async function produceMetadata(items: Dynamo[], producer: Producer) {
+  try {
+    await init();
+  } catch (initError) {
+    console.error("Error initializing coin metadata cache:", initError);
+  }
+
+  const metadataMessages: string[] = [];
+
+  items.map((item) => {
+    const record = getMetadataRecord(item);
+    if (!record) return;
+    validate(record, "coins-metadata");
+
+    const msg = JSON.stringify(record);
+    if (!metadataMessages.includes(msg)) {
+      metadataMessages.push(msg);
+    }
+  });
+
+  if (process.env.KAFKA_UP === "true") {
+    try {
+      await producer.send({
+        topic: "coins-metadata",
+        messages: metadataMessages.map((value) => ({ value })),
+      });
+      return;
+    } catch (error) {
+      console.error('Kafka failed for "coins-metadata". Fallback to ES.', error);
+    }
+  }
+
+  try {
+    const body: any[] = [];
+    for (const msg of metadataMessages) {
+      const doc = JSON.parse(msg);
+      body.push({ index: { _index: "coins-metadata" } });
+      body.push(doc);
+    }
+    if (body.length) {
+      await esClient.bulk({ body });
+    }
+  } catch (err) {
+    console.error("Metadata fallback (bulk to ES) failed:", err);
+  }
+}
+
+async function produceTopics(items: Dynamo[], topic: Topic, producer: Producer) {
+  const messages: string[] = [];
+
+  items.map((item) => {
+    const { price, SK } = item;
+
+    if (topic !== "coins-timeseries" && SK !== 0) return;
```

Comment:
can you change it to two explicit checks, timeseries & SK = 0 means return

And coins-current & SK != 0 means return


More verbose, but logic is more clear for other llamas if they stumble upon this code